File: programmers/level1/키패드 누르기.py
def solution(numbers, hand):
    def get_distance(lt, rt, n):
        ld = abs((lt-1)//3 - (n-1)//3)
        rd = abs((rt-1)//3 - (n-1)//3)
        if lt in [1,4,7,10]: ld += 1
        if rt in [3,6,9,12]: rd += 1
        return ld, rd
    
    ans, lt, rt = "", 10, 12
    
    for n in numbers:
        if n==0: n=11
        if n in [1,4,7]:
            ans += "L"
            lt = n
        elif n in [3,6,9]:
            ans += "R"
            rt = n
        else:
            ld, rd = get_distance(lt, rt, n)
            if (ld==rd and hand=="left") or ld<rd:
                ans += "L"
                lt = n
            else:
                ans += "R"
                rt = n
    return ans

File: programmers/level1/키패드 누르기.java
class Solution {
    public String solution(int[] numbers, String hand) {
        StringBuilder sb = new StringBuilder();
        int nowLeft = 9;
        int nowRight = 11;
        for(int n: numbers) {
            if(n==0) n = 10;
            else n = n-1;
            if(n==0 || n==3 || n==6) {
                sb.append("L");
                nowLeft = n;
            } else if(n==2 || n==5 || n==8) {
                sb.append("R");
                nowRight = n;
            } else if(getLDistance(nowLeft, n) < getRDistance(nowRight, n)) {
                sb.append("L");
                nowLeft = n;
            } else if(getLDistance(nowLeft, n) > getRDistance(nowRight, n)) {
                sb.append("R");
                nowRight = n;
            } else if(hand.equals("left")) {
                sb.append("L");
                nowLeft = n;
            } else if(hand.equals("right")) {
                sb.append("R");
                nowRight = n;
            }
        }
        return sb.toString();
    }
    
    public int getLDistance(int nowLeft, int n) {
        int ret = Math.abs(nowLeft/3 - n/3);
        if(nowLeft == 1 || nowLeft==4 || nowLeft==7 || nowLeft==10) {
            return ret;
        } else {
            return ret+1;
        }
    }
    
    public int getRDistance(int nowRight, int n) {
        int ret = Math.abs(nowRight/3 - n/3);
        if(nowRight == 1 || nowRight==4 || nowRight==7 || nowRight==10) {
            return ret;
        } else {
            return ret+1;
        }
    }
}